metalearn/metafeatures/base.py

```diff
@@ -0,0 +1,146 @@
+            the same order as ``computer`` returns them.
+        problem_type
+            The type of ML problem `computer`'s metafeatures can be computed for.
+        groups
+            The metafeature groups this computer's returned metafeatures belong to.
+            e.g. statistical, info-theoretic, simple, etc.
+        argmap
+            A custom map of ``computer``'s argument names to the global resource names
+            that will be passed as ``computer``'s arguments when ``computer`` is called.
+        """
+        super(MetafeatureComputer, self).__init__(computer, returns, argmap)
+        self.groups = groups
+        self.problem_type = problem_type
+
+
+class ResourceComputerMap(MutableMapping):
+    def __init__(self, computers: Union[ResourceComputer,List[ResourceComputer],None] = None) -> None:
+        """
+        Wraps a dictionary map of resource names to their computers.
+        Includes visibility into whether duplicate computers
+        are trying to become associated with a resource in the map e.g.
+        if a package developer has accidentally declared two computers
+        that return the same resource.
+        """
+        self._map: Dict[str,ResourceComputer] = {}
+        if computers is not None:
+            self.add(computers)
+    
+    def add(self, computers: Union[ResourceComputer,List[ResourceComputer]]) -> None:
+        """
+        Adds more resource name/resource computer key/value
+        pairs to a resource map, throwing an error on duplicates.
+        """
+        if isinstance(computers, list):
+            for computer in computers:
+                self._add_one(computer)
+        elif isinstance(computers, ResourceComputer):
+            self._add_one(computers)
+        else:
+            raise ValueError("computers must be ResourceComputer or List[ResourceComputer]")
+    
+    def __getitem__(self, key: str = None) -> ResourceComputer:
+        """Used for getting a resource from the map."""
+        return self._map[key]
+
+    def _add_one(self, computer: ResourceComputer) -> None:
+        if not isinstance(computer, ResourceComputer):
+            raise ValueError(f"computer is not a ResourceComputer; it is a {type(computer)}")
+            
+        for resource_name in computer.returns:
+            self.__setitem__(resource_name, computer)
+    
+    def __setitem__(self, resource_name: str, computer: ResourceComputer):
+        if resource_name in self._map:
+            raise ValueError(
+                f"duplicate computer '{computer.name}' provided for resource '{resource_name}', "
+                f"which is already present in the resouce map, registered "
+                f"by computer '{self._map[resource_name].name}'"
+            )
+        self._map[resource_name] = computer
+    
+    def __iter__(self):
+        return iter(self._map)
+    
+    def __len__(self):
+        return len(self._map)
+    
+    def __delitem__(self, key: str):
```

Comment:
I am not sure this is a concern for us...

Reply:
An implementation of this method is required when subclassing `MutableMapping`, since `MutableMapping` declares `__delitem__` as one of its abstract methods.